### PR TITLE
Add desi_proc_tilenight to desi_run_night

### PR DESCRIPTION
**Pull request by marcelo-alvarez**
Adds the option (by providing the --tilenight flag) to use desi_proc_tilenight in desi_run_night. 

Testing desi_run_night with this branch against the current master branch for night 20211014 on Perlmutter:
```
switch_to_repo(){
  repo=$1
  branch=$2
  branchdir=$SCRATCH/$repo
  git clone -b $branch git@github.com:desihub/$repo.git $branchdir
  export PYTHONPATH=$branchdir/py:$PYTHONPATH 
  export PATH=$branchdir/bin:$PATH    
} 

source /global/common/software/desi/desi_environment.sh main
DESI_SPECTRO_REDUX=/global/cfs/cdirs/desi/users/$USER/spectro/redux

SPECPROD=ref-rnight-gpuext-gpuspc
desi_run_night -n 20211014 --z-submit-types cumulative -q regular --system-name perlmutter-gpu  

switch_to_repo desispec desi_run_tnight
SPECPROD=rnight-gpuext-gpuspc
desi_run_night -n 20211014 --z-submit-types cumulative -q regular --system-name perlmutter-gpu --tilenight
```
results in identical (arc, ccdcalib, flat, nightlyflat, and coadd)-prefixed batch Slurm files.

Additionally, 23 tilenight-prefixed batch Slurm files are produced when --tilenight is provided as an argument on this branch, corresponding to the 23 tiles for that night, instead of the 34 prestdstar, 34 poststdstar, and 23 stdstar batch Slurm files for the master branch.

All processing for the --tilenight run of desi_run_night seems to have completed successfully, and comparison of outputs to a run from the master branch is ongoing. 

@akremin and @sbailey please go ahead and have a look at the structure of the changes and test it out.

### CODE REVIEW

**Review by sbailey — COMMENTED** [review]
Thanks for this important update.  I haven't tried running it myself, but from inspecting your test production it appears to have worked, including the tilenight entries in the processing table.

I put one question/comment inline about if/elif/else flow for tilenight vs. not.

Variable naming consistency: in `scripts.submit_night` the variable `tilenight` is a boolean, while in `workflow.procfuncs` it is a function, and if I saw that variable name out of context I would have guessed it was a tuple `(tile, night)`.  Please update the boolean case to be something like `use_tilenight` and the function name to be something like `submit_tilenight` (and matching `submit_redshifts` for the newly added `redshifts` function).

It appears that the flat*.slurm jobs are not using GPUs and gpu_specter, even though they also do extractions.  I think we want to update those in this PR to also use GPUs.

FTR: Other jobs that don't use the GPUs but are still submitted to the GPU partition:
ccdcalib, arc, psfnight, nightlyflat.  I think that's ok for now, and a sparate PR could add functionality for submitting perlmutter jobs with some going to CPUs and others going to GPUs.

Let's also get @akremin review of this since it touches a lot of the processing table details that he originally authored.

File: py/desispec/scripts/submit_night.py
Comment (marcelo-alvarez) [reply py/desispec/scripts/submit_night.py]: > Please clarify in comments what each does and how they are different, and consider whether some of the if/elif/else branching logic can be updated to make this more explicit.

I've changed the logic to make it clearer; if done with science exposures for a tile and use_tilenight==True, then `submit_tilenight_and_redshifts` is called, otherwise `checkfor_and_submit_joint_job` is called, as before.

**Comment by marcelo-alvarez** [comment]
@sbailey thanks for your comments and review. See above inline comment about the clarity of the logic, which has been improved. I also replaced `--tilenight` with `--use-tilenight` and the functions `tilenight` and `redshifts` with `submit_tilenight` and `submit_redshifts`.

Note that running without `--use-tilenight` on Perlmutter now results in flats running on a single node, using either 64 or 128 tasks for gpu and cpu, respectively. 

Some timings for night 20211014 with different systems / options:
```
system tn    flats     arcs     sci   tot     tot 
            (n=12)    (n=5)  (n=34)        actual

 c-knl  N   4.4(3)  2.7(10) 14.6(3)  21.7   108.7
 c-hsw  N   3.6(2)  2.0( 3) 10.8(2)  16.5    47.2

 p-cpu  N   2.4(2)  1.6( 3)  9.8(2)  14.0    14.0
 p-cpu  Y   1.7(1)  1.5( 3)  6.0(1)   9.3     9.3
 
 p-gpu  Y   2.1(1)  2.3( 5)  5.5(1)  10.0    10.0

number of nodes used in step in parentheses 
sci = tilenight OR prestdstar + stdstar + poststdstar
tn = with --use-tilenight
knl node-hour charge factor:    0.2
hsw node-hour charge factor:    0.35
p-gpu and p-cpu charge factors: 1.0
```
On a node-for-node wall-clock basis, we see a speed-up of about a factor of 5 and 10 with respect to cori-haswell and cori-knl, respectively, for either perlutter-gpu or perlmutter-cpu.

@akremin please have a look and make sure it runs as you expect for cori before we merge.

**Review by sbailey — APPROVED** [review]

**Comment by sbailey** [comment]
Thanks.  This looks good to me but let's wait for a cross check from @akremin on the processing table logic before merging.

**Comment by marcelo-alvarez** [comment]
An additional comment on performance and speedup, though addressing this should *not* delay the merging of these changes, as they are, significantly.

Note above that there is a negligible speed up achieved for science exposures using tilenight on GPUs vs CPUs (5.9 vs 6.0 node-hours for processing 34 exposures through the prestdstar, stdstar, and poststdstar steps). Here is a comparison of the total time of individual steps in tilenight for a single tile, (tileid=6699, see
```
/global/cfs/cdirs/desi/users/malvarez/spectro/redux/drt1-cpu-1802/run/scripts/night/20211014/tilenight-20211014-6699-2442922.log
/global/cfs/cdirs/desi/users/malvarez/spectro/redux/drt1-gpu-1802/run/scripts/night/20211014/tilenight-20211014-6699-2437723.log
```
), using GPUs and CPUs, columns 2, 3, and 4 show time per exposure (3 in the case of tile 6699) in seconds:
```
                         name  drt1-cpu-1802  drt1-gpu-1802  ratio  difference

                        total          554.3          552.0   1.00        -2.3

                      extract          357.6          253.5   0.71      -104.1
                      imports           50.9           19.0   0.37       -32.0
fiberflat_humidity_correction           16.7           89.8   5.37        73.1
                   traceshift           13.0           25.5   1.96        12.5
                  exposure_qa           26.3           41.6   1.58        15.3
                    fluxcalib           19.3           27.5   1.42         8.2
                       skysub           14.9           22.6   1.51         7.7
                    preflight            2.5            8.2   3.30         5.7
                     preproc           37.2           41.8   1.12         4.6
                      picksky            0.9            3.5   3.83         2.6
                     fibermap            1.7            4.1   2.38         2.4
                   applycalib           12.1           14.4   1.19         2.3
```
The extractions are ~30% faster (~100 seconds per exposure) using GPUs, as expected, but the fiberflat_humidity_correction step takes more than 4 times longer (~70 seconds per exposure) using GPUs, canceling out most of the gains achieved for the extractions. Note also that for some reason imports took significantly longer for CPUs, and this makes the GPU numbers look better than they would be if importing didn't take so long. Other, non GPU, steps take longer on GPU nodes, which is consistent with having half the number of processors available, to the extent that these steps are making efficient use of the extra cores. These trends are observed for other tiles as well for this pair of runs. 

Mentioning @dmargala in case he has any comment on this from his scale run experience. It may make sense to see what can be done to minimize additional overhead in a separate issue / pull request.

File: py/desispec/workflow/procfuncs.py
Comment (akremin) [inline py/desispec/workflow/procfuncs.py:1030]: Can you specify if it's a processing table row or an exposure table row? I believe it's a processing row (prow),

File: py/desispec/workflow/procfuncs.py
Comment (akremin) [inline py/desispec/workflow/procfuncs.py:1134]: Would it be good to verify that all exposures have the same TILEID?
This is a lower level function but it may be a nice safety mechanism.

**Review by akremin — CHANGES_REQUESTED** [review]
This is a big step in the right direction and the changes look very nice. I asked a few questions in-line and requested a few changes/clarifications. Nothing is a major issue but there appear to be a couple corner-cases that we should patch up. Let me know if I've misread anything, I'm happy to iterate on things.

**Comment by marcelo-alvarez** [comment]
> This is a big step in the right direction and the changes look very nice. I asked a few questions in-line and requested a few changes/clarifications. Nothing is a major issue but there appear to be a couple corner-cases that we should patch up. Let me know if I've misread anything, I'm happy to iterate on things.

Thanks @akremin for the detailed review. I will address your questions in-line.

**Comment by marcelo-alvarez** [comment]
@akremin @sbailey I just pushed changes that avoid timeouts resulting from too short a time limit for tilenight jobs on multiple exposures. Time factors are now changed to:
```
perlmutter-cpu:
    site: NERSC
...
    timefactor: 0.65

perlmutter-gpu:
    site: NERSC
...
    timefactor: 0.5
```
and the scaling for tilenights is 
```
    elif jobdesc == 'TILENIGHT':
        ncores, nodes = config['cores_per_node'], 1
        # ~150 frames per node hour plus
        # 8-minute overhead for a single node
        # perlmutter-[cpu,gpu] have timefactor=[0.7,0.5]
        runtime = 8 + int(60. / 140. * ncameras * nexps)
```
The attached figure shows a scatter plot of actual run tilenight runs with different numbers of exposures for nights 20211014, 20210203, and 20220508, along with the scaling and timefactor values above. The "padding" in time is my attempt to compromise between minimizing non-system-related timeouts and having jobs wait on the queue for longer than necessary due to their time limit. 

tilenight-scaling

File: py/desispec/scripts/submit_night.py
Comment (marcelo-alvarez) [reply py/desispec/scripts/submit_night.py:393]: @akremin prow corresponds to the new exposure, since it is defined in line 368:
```
prow = define_and_assign_dependency(prow, calibjobs)
```
and that statement is executed regardless of whether `use_tilenight==True` or not. So I think I'm appending a new exposure here in the case `use_tilenight==True`. 

Please let me know if this makes sense, and I will mark this conversation as resolved.

File: py/desispec/workflow/procfuncs.py
Comment (marcelo-alvarez) [reply py/desispec/workflow/procfuncs.py:508]: > Also, this won't define a `JOBDESC` if `use_tilenight` is set. Don't you want it to assign `JOBDESC` to be `"tilenight"`?

The statement 
```
joint_prow['JOBDESC'] = 'tilenight'
``` 
in `make_tnight_prow` already assigns `JOBDESC` to `"tilenight"` before `define_and_assign_dependency` is called.

File: py/desispec/workflow/procfuncs.py
Comment (marcelo-alvarez) [reply py/desispec/workflow/procfuncs.py:1134]: @akremin verifying that all exposures in a tilenight job have the same TILEID is already taken care of at a lower level in `desi_proc_tilenight` itself; the information about exposures here is only used for setting the time limit.

File: py/desispec/workflow/desi_proc_funcs.py
Comment (marcelo-alvarez) [reply py/desispec/workflow/desi_proc_funcs.py:494]: Done.

File: py/desispec/scripts/submit_night.py
Comment (marcelo-alvarez) [reply py/desispec/scripts/submit_night.py:328]: @akremin thanks for catching this. I added the lines 
```
    sciexps = (etable['OBSTYPE']=='science')
    scisrtd = etable[sciexps].argsort(['TILEID'])
    etable[sciexps] = etable[sciexps][scisrtd]
```
above to do this.

File: py/desispec/scripts/submit_night.py
Comment (marcelo-alvarez) [reply py/desispec/scripts/submit_night.py]: @akremin thanks for catching this. I replaced
```
        if use_tilenight:
```
with
```
        if use_tilenight and len(sciences)>0:
```
so the last set of calibrations should now always be submitted for joint-processing for nights with just calibrations, even when `use_tilenight` is set.

File: py/desispec/workflow/procfuncs.py
Comment (marcelo-alvarez) [reply py/desispec/workflow/procfuncs.py:354]: @akremin thanks for catching this. Since I designed `create_desi_proc_tilenight_batch_script` in such a way that the code to generate the `desi_proc_tilenight` command for batch scripts is local to that function, it doesn't make sense to report that command here. I changed 
```
            log.info("Command to be run: {}".format(cmd.split()))
```
to
```
            if prow['JOBDESC'] != 'tilenight':
                log.info("Command to be run: {}".format(cmd.split()))
```
to avoid the crashing you mentioned.

File: py/desispec/workflow/procfuncs.py
Comment (marcelo-alvarez) [reply py/desispec/workflow/procfuncs.py:509]: > Please update the docstring.

Done.

File: py/desispec/workflow/procfuncs.py
Comment (marcelo-alvarez) [reply py/desispec/workflow/procfuncs.py]: Done.

File: py/desispec/scripts/submit_night.py
Comment (sbailey) [reply py/desispec/scripts/submit_night.py:401]: @marcelo-alvarez can you clarify: on line 368 (line numbers from this PR) `prow` is created regardless of `use_tilenight` True or False., but where is the equivalent logic for the block under line 369:
```
if (not use_tilenight) or erow['OBSTYPE'] != 'science':
    ...
    prow = create_and_submit(prow,  ...)
    ....
    ptable.add_row(prow)
    ...
```
for the case `use_tilenight==True and erow['OBSTYPE'] == 'science'`?

Comment (marcelo-alvarez) [reply py/desispec/scripts/submit_night.py:401]: @sbailey when `use_tilenight=True` there are no processing steps for individual science exposures, so there are no science exposure `prow`s to append, and hence there is no equivalent logic for the block under line 369, the purpose of which is to create and submit per-exposure steps and append them to the processing table.

Rather, the `prow` created at line 368 (for the version you reference above) 
```
        prow = define_and_assign_dependency(prow, calibjobs)
```
when `use_tilenight=True` is only used to append to the list of science exposures, i.e. 
```
        elif curtype == 'science' and (prow['LASTSTEP'] != 'skysub' or use_tilenight):
            sciences.append(prow)
```

**Comment by sbailey** [comment]
Thanks.  I believe all of @akremin's requested changes have been addressed except for the logging from --dry-run-level 2 which we can punt to the future.  I just addressed the merge conflicts in the perlmutter job runtime config and will merge after tests re-pass.

File: py/desispec/workflow/procfuncs.py
Comment (marcelo-alvarez) [reply py/desispec/workflow/procfuncs.py:354]: @akremin @sbailey to clarify; printing the command line(s) in the batch file for `dry_run>1` is a desirable feature that tilenight batch scripts do not currently have, which was requested above. It is not clear to me how to do this in a consistent way without creating a new function whose only purpose is to create the command line for `desi_proc_tilenight` arguments, which I don't think is a good idea. Let's address how to do this after merging this PR.

**Pull request merged**
